pkg/providers/azure/core/provider.go

```diff
@@ -0,0 +1,75 @@
+package core
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+)
+
+type AzProvider interface {
+	SubscriptionId() string
+	ResourceGroupName() string
+	ServicePrincipalToken(resource string) (*adal.ServicePrincipalToken, error)
+}
+
+type azProviderImpl struct {
+	env    auth.EnvironmentSettings
+	subId  string
+	rgName string
+}
+
+func (p *azProviderImpl) ServicePrincipalToken(resource string) (*adal.ServicePrincipalToken, error) {
+	if fileCred, fileErr := auth.GetSettingsFromFile(); fileErr == nil {
+		return fileCred.ServicePrincipalTokenFromClientCredentialsWithResource(resource)
+	} else if clientCred, clientErr := p.env.GetClientCredentials(); clientErr == nil {
+		clientCred.Resource = resource
+		return clientCred.ServicePrincipalToken()
+	} else if clientCert, certErr := p.env.GetClientCertificate(); certErr == nil {
+		clientCert.Resource = resource
+		return clientCert.ServicePrincipalToken()
+	} else if userPass, userErr := p.env.GetUsernamePassword(); userErr == nil {
+		userPass.Resource = resource
+		return userPass.ServicePrincipalToken()
+	} else {
+		fmt.Printf("error retrieving credentials:\n -> %v\n -> %v\n -> %v\n -> %v\n", fileErr, clientErr, certErr, userErr)
+	}
+
+	msiConf := p.env.GetMSI()
+	msiConf.Resource = resource
+	return msiConf.ServicePrincipalToken()
+}
+
+func (p *azProviderImpl) SubscriptionId() string {
+	return p.subId
+}
+
+func (p *azProviderImpl) ResourceGroupName() string {
+	return p.rgName
+}
+
+var _ AzProvider = &azProviderImpl{}
+
+func New() (*azProviderImpl, error) {
```

Comment:
this might also be a nice lib to use: https://github.com/kelseyhightower/envconfig/

Reply:
Let's look at this as a broader refactor, I also wanted to look at using https://github.com/google/wire to build our membrane provider implementation instead of hand crafting them as well.